src/Features/Core/Portable/DisposeAnalysis/DisposableFieldsShouldBeDisposedDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,266 @@
+
+                // If we have any potential disposable object creation descendant within the operation blocks,
+                // register an operation action to analyze field references where field might be assigned a disposable object.
+                if (_disposeAnalysisHelper.HasAnyDisposableCreationDescendant(operationBlockStartContext.OperationBlocks, containingMethod))
+                {
+                    operationBlockStartContext.RegisterOperationAction(AnalyzeFieldReference, OperationKind.FieldReference);
+                }
+
+                // If this is a Dispose method, then analyze dispose invocations for fields within this method.
+                if (containingMethod.GetDisposeMethodKind(_disposeAnalysisHelper.IDisposable, _disposeAnalysisHelper.Task) != DisposeMethodKind.None)
+                {
+                    AnalyzeDisposeMethod();
+                }
+
+                return;
+
+                // Local function
+                void AnalyzeFieldReference(OperationAnalysisContext operationContext)
+                {
+                    var fieldReference = (IFieldReferenceOperation)operationContext.Operation;
+                    var field = fieldReference.Field;
+
+                    // Check if this is a Disposable field that is not currently being tracked.
+                    if (_fieldDisposeValueMap.ContainsKey(field) ||
+                        !_disposableFields.Contains(field))
+                    {
+                        return;
+                    }
+
+                    // Only track instance fields on the current instance.
+                    if (field.IsStatic || fieldReference.Instance?.Kind != OperationKind.InstanceReference)
+                    {
+                        return;
+                    }
+
+                    // We have a field reference for a disposable field.
+                    // Check if it is being assigned a locally created disposable object.
+                    if (fieldReference.Parent is ISimpleAssignmentOperation simpleAssignmentOperation &&
+                        simpleAssignmentOperation.Target == fieldReference)
+                    {
+                        if (lazyPointsToAnalysisResult == null)
+                        {
+                            if (_disposeAnalysisHelper.TryGetOrComputeResult(operationBlockStartContext.OperationBlocks,
+                                containingMethod, operationBlockStartContext.Options, s_disposableFieldsShouldBeDisposedRule, trackInstanceFields: false,
+                                trackExceptionPaths: false, operationBlockStartContext.CancellationToken,
+                                out _, out var pointsToAnalysisResult) &&
+                                pointsToAnalysisResult != null)
+                            {
+                                Interlocked.CompareExchange(ref lazyPointsToAnalysisResult, pointsToAnalysisResult, null);
```

Comment:
is this actually ever run concurrently?, lazyPointsToAnalysisResult is local variable shared by local function which given to OperationBlockStartContext (operationBlockStartContext.RegisterOperationAction(AnalyzeFieldReference, OperationKind.FieldReference);) 

do we actually run action inside of BLockStartContext concurrently? I thought BlockContexts run concurrently not ones registered to BLockContext itself.

and since lazyxxx is local variable, those won't be shared between different actions?

Reply:
This variable is shared between multiple callbacks to `AnalyzeFieldReference` within the same method. Current driver implementation calls them sequentially, but that is implementation detail which might change in future :)